src/libcharon/plugins/kernel_super/kernel_super_ipsec.c

```diff
@@ -2262,6 +2263,34 @@ static u_int tunnel_counter_key_hash(const void *key) {
 						  chunk_hash(chunk_from_str((char*)tunnel_key->vrf_name)))));
 }
 
+static struct tunnel_counter_key *create_new_tunnel_counter_key(uint32_t terminator_id,
+																const char* tenant_id,
+																const char* account_id,
+																const char* vrf_name){
+	size_t tenant_id_len = strlen(tenant_id);
+	size_t account_id_len = strlen(account_id);
+	size_t vrf_name_len = strlen(vrf_name);
+	struct tunnel_counter_key *key = (struct tunnel_counter_key *)malloc(sizeof(struct tunnel_counter_key) +
+																		tenant_id_len + 1 +
+																		account_id_len + 1 +
+																		vrf_name_len + 1);
+	key->terminator_id = terminator_id;
+	key->tenant_id = (char*)(key + sizeof(struct tunnel_counter_key));
+	key->account_id = key->tenant_id + tenant_id_len + 1;
+	key->vrf_name = key->account_id + account_id_len + 1;
```

Comment:
https://en.wikipedia.org/wiki/KISS_principle

There is no reason to have such complication for such a simple task. It is far far more readable and less error prone to store in `tunnel_counter_key` buffers with maximum size as in `struct super_cert_info` instead of doing this dance with pointers.. (for example `key->tenant_id` may point outside the struct since `key + sizeof(struct tunnel_counter_key)` adds `sizeof(struct tunnel_counter_key) * sizeof(struct tunnel_counter_key)` bytes to the address `key` points to. It should be `key + 1`.

Another simple alternative is to have these strings live in the heap as well as the key, though I prefer using maximum-size buffers IMO.
But all in all this handling with pointers is unnecessarily complicated